nomad/state/state_store.go

```diff
@@ -230,18 +230,18 @@ func (s *StateStore) UpsertPlanResults(index uint64, results *structs.ApplyPlanR
 		return err
 	}
 
```

Comment:
Looks like we used to update the Job field on all allocs (including stopped) in 0.9.1. The plan applier used to add both stopped and new allocs here [1] , and the state store would set the job from the plan to all allocs [2]. 

So something else must have changed in tandem to cause this panic in 0.9.2. Were you able to determine if any other additional changes besides the plan applier changes contributed to this? 

 [1] https://github.com/hashicorp/nomad/blob/v0.9.1/nomad/plan_apply.go#L168  
 [2] https://github.com/hashicorp/nomad/blob/v0.9.1/nomad/state/state_store.go#L191

Was there something in allocrunner that changed as well in 0.9.2?

Reply:
In 0.9.1, `DenormalizeAllocationJobs` would set `alloc.Job` if it's not set already and it's not terminal: https://github.com/hashicorp/nomad/blob/v0.9.1/nomad/structs/funcs.go#L275-L286  .

I confirmed this behavior by adding some logging statements in various places.  I can demo it offline.